### DISCUSSION TITLE
Content Items event handler: how to intercept the saving event of a Content Item?

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15568] [Q&A] [answered]
Hi everyone,

I'm trying to understand if there's the possibility of using an **event handler to intercept when a Content Item, created in the backoffice, is saved by the admin**. I know there's the possibility to use `ContentPartHandler<TPart>` and `ContentFieldHandler<TField>`, as well explained in this [video](https://youtu.be/Wjizx4XU2V4). But how about Content Items?

**I need to intercept the saving of a `Book` Content Item, to perform some checks on a `BagPart` contained within the `Book`.** How can I intercept the Content Item saving event?

I imagine something like this:

```csharp
using OrchardCore.ContentManagement.Handlers;
using OrchardCore.ContentManagement;
using System.Threading.Tasks;

public class CustomContentItemHandler : ContentHandler
{
    public override async Task UpdatingAsync(UpdateContentContext context)
    {
        var contentItem = context.ContentItem;

        // Intercept Book Content Item
        if (contentItem.ContentType == "Book")
        {
            // Retrieve here the "Bag" part contained in it, and perform checks
        }

        await base.UpdatingAsync(context);
    }
}
```

1. Of course, a `ContentHandler` like this one doesn't exist, but is that possible, or am I dreaming? 😁
2. Or is there a way to intercept the save event for a non-custom `ContentPart` or `ContentField`, such as the `BagPart`, already available in Orchard Core, **and contained in a specific Content Item**?

Thank you

**Comment by douwinga** [comment]
You are looking for `ContentHandlerBase`. Personally, unless this content type is created via a migration in code and also has an associated `BookPart` with model in the same module, I would just handle something like this in a workflow using the Content Updated event. And if you had a BookPart you would use the `ContentPartHandler` anyways. Targeting a specific content type in a `ContentHandlerBase` just feels like the wrong tool to me.

**Reply by douwinga** [reply, marked as answer]
I think I missed your second question. Yes you can have event handlers on existing parts such as `BagPart`. Just do the same thing you would for a custom part `public class MyBagPartHandler : ContentPartHandler<BagPart>` and then in the startup `services.AddContentPart<BagPart>().AddHandler<MyBagPartHandler>();`. It may seem odd to be calling `AddContentPart` on a Part that has already been added, but if you look, this method calls `GetOrAddContentPart`, so you are not adding it twice.

Now that I think I understand what you are trying to do, this is probably the best approach and you can ignore the workflow idea. ContentPartHandlers can get the content item easily enough, so for some reason if you had to do a check for the content type of `Book` you could still do that.

**Reply by MarGraz** [reply]
@douwinga thank you for your reply, as always 😊

Essentially, **I need to intercept the saving of a `Book`** - a content item created from the backoffice UI, not a custom one -. **I need to take the `BagPart` within it**, and checking the number of items in the `BagPart` against the `Min` and `Max` values - we discussed it [here](https://github.com/OrchardCMS/OrchardCore/discussions/15549#discussioncomment-8845326) -. If the items do not meet these criteria, I want to send an error message back to the admin.

I will try using the `ContentPartHandler` approach. Let's see how it turns out.

Thank you

**Reply by MarGraz** [reply]
@douwinga it works 🥳
I also added a check on the Content Type `Book`, **because this `BagPart` handler is called every time a Content Item has a `BagPart` in it.**

**I'm not sure if I need to add a null check on `Result` property**, in `var partDefinition = contentTypeDefinition.Result{here}`.
This is an example of what I have implemented:

```csharp
using CodingOps.Custom.ContentFields.Models;
using OrchardCore.ContentManagement.Handlers;
using OrchardCore.ContentManagement.Metadata;
using OrchardCore.Flows.Models;

public class BookBagPartHandler : ContentPartHandler<BagPart>
{
    // ContentDefinitionManager is necessary to retrieve the custom settings
    // ("max" and "min" values) I have added to every BagPart.
    // See this discussion: https://github.com/OrchardCMS/OrchardCore/discussions/15549#discussioncomment-8845326
    private readonly IContentDefinitionManager _contentDefinitionManager;

    public BookBagPartHandler(IContentDefinitionManager contentDefinitionManager)
    {
        _contentDefinitionManager = contentDefinitionManager;
    }

    public override async Task UpdatingAsync(UpdateContentContext context, BagPart part)
    {
        // Check if the type of the BagPart is a Book Content Item
        if (context.ContentItem.ContentType == "Book")
        {
            // Retrieve the settings
            var contentTypeDefinition = _contentDefinitionManager.GetTypeDefinitionAsync("Book");
            var partDefinition = contentTypeDefinition.Result.Parts.FirstOrDefault(p => p.PartDefinition.Name == nameof(BagPart));
            var settings = partDefinition.GetSettings<CustomBagPartSettings>();

            int minItems = settings.MinItems;
            int maxItems = settings.MaxItems;

            // Now I can use minItems and maxItems as needed
            int itemCount = part.ContentItems.Count;
            if (itemCount < minItems || itemCount > maxItems)
            {
                // Perform my validation or other logic here
            }
        }

        await Task.CompletedTask;
    }
}
```

Thank you

**Reply by douwinga** [reply]
typically you would use `await` instead of using `.Result` because a) it is cleaner looking code and b) .Result can end up causing deadlocks. Something like this:

```csharp
public override async Task UpdatedAsync(UpdateContentContext context, BagPart part)
{
    // Check if the type of the BagPart is a Book Content Item
    if (context.ContentItem.ContentType != "Book")
    {
        return;
    }

    // Retrieve the settings
    var contentTypeDefinition = await _contentDefinitionManager.GetTypeDefinitionAsync("Book");
    var partDefinition = contentTypeDefinition?.Parts?.FirstOrDefault(p => p.PartDefinition.Name == nameof(BagPart));
    var settings = partDefinition?.GetSettings<CustomBagPartSettings>();

    if (settings == null)
    {
        return;
    }

    int minItems = settings.MinItems;
    int maxItems = settings.MaxItems;

    // Now I can use minItems and maxItems as needed
    int itemCount = part.ContentItems.Count;
    if (itemCount < minItems || itemCount > maxItems)
    {
        // Perform my validation or other logic here
    }
}
```

**Reply by MarGraz** [reply]
@douwinga **in my previous comment I changed the `UpdatedAsync` method in `UpdatingAsync`**, because I need to perform the validation before the saving. 

I saw that there's no `context.fail()` method for the `UpdateContentContext` type. **How can I return an error message to the backoffice UI in case the validation fails?**

Thank you

**Reply by douwinga** [reply]
Should be able to just inject `IActionContextAccessor` into your handler and then use that to add a model state error I think. `_actionContextAccessor.ActionContext?.ModelState?.AddModelError("YourKey", "Your error message");`. Having a model state error should cause the session to cancel causing a "rollback". I would look at the `ValidatingAsync` event though and make sure that doesn't make more sense to use. I have never used them so not sure if this use case is their purpose.

**Reply by MarGraz** [reply]
@douwinga **It works**, the process successfully performs the rollback and displays the error message.

I also tried to use both `ValidatingAsync()` and `ValidatedAsync()` methods. The `ValidateContentPartContext` class, provided as a parameter, allows the use of the `context.Fail()` method. However, **these methods are never invoked when I save my `Book` content item**, leaving me confused about their intended use cases 🤔 
Also the `IContentPartHandler` interface ([here](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/IContentPartHandler.cs)) doesn't have any `summary` for each method, which could have offered some guidance 😟

Thank you for your help
